Add jsblocks

astoilkov opened the pull request:
[jsblocks](http://jsblocks.com) have been developed for the past couple of years. Two weeks ago it was released and collected more than 800 stars. You could take a look at some of the advantages of the project [here](http://jsblocks.com/learn/introduction-why-jsblocks). Some of the things include:
* Performance
* Debugging exprience
* Server-side rendering

Test results:

```
TodoMVC - blocks
  When page is initially opened
    √ should focus on the todo input field
  No Todos
    √ should hide #main and #footer (109ms)
  New Todo
    √ should allow me to add todo items (459ms)
    √ should clear text input field when an item is added (288ms)
    √ should append new items to the bottom of the list (570ms)
    √ should trim text input (262ms)
    √ should show #main and #footer when items added (253ms)
  Mark all as completed
    √ should allow me to mark all items as completed (610ms)
    √ should allow me to clear the completion state of all items (687ms)
    √ complete all checkbox should update state when items are completed / cleared (739ms)
  Item
    √ should allow me to mark items as complete (575ms)
    √ should allow me to un-mark items as complete (585ms)
    √ should allow me to edit an item (942ms)
    √ should show the remove button on hover
  Editing
    √ should hide other controls when editing (672ms)
    √ should save edits on enter (867ms)
    √ should save edits on blur (942ms)
    √ should trim entered text (868ms)
    √ should remove the item if an empty text string was entered (800ms)
    √ should cancel edits on escape (939ms)
  Counter
    √ should display the current number of todo items (399ms)
  Clear completed button
    √ should display the correct text (571ms)
    √ should remove completed items when clicked (865ms)
    √ should be hidden when there are no items that are completed (615ms)
  Persistence
    √ should persist its data (854ms)
  Routing
    √ should allow me to display active items (633ms)
    √ should respect the back button (892ms)
    √ should allow me to display completed items (616ms)
    √ should allow me to display all items (776ms)
    √ should highlight the currently applied filter (678ms)


30 passing (3m)
```

mau-azevedo commented:
This is a good Idea to put this on TodoMVC, I test JSBlock and like of your performance.

arthurvr commented on examples/blocks/.editorconfig:
```diff
@@ -0,0 +1,15 @@
+indent_size = 2
+
+[*.md]
+trim_trailing_whitespace = false
```
Please remove this file. We have our own code styling rules and those are in the root directory :) Same thing for `.gitattributes` :)

arthurvr commented on examples/blocks/.gitignore:
```diff
@@ -0,0 +1,3 @@
+# Folders
+.idea
+/node_modules
```
Let's remove the comment and `.idea`. That should be in your global gitignore. Please build a gitignore to easily manage your dependencies [as other examples](https://github.com/tastejs/todomvc/blob/master/examples/angularjs/.gitignore).

arthurvr commented on examples/blocks/index.html:
```diff
@@ -0,0 +1,58 @@
+<!doctype html>
+<html lang="en">
```
misses a `data-framework` attribute such as the other examples.

arthurvr commented on examples/blocks/index.html:
```diff
@@ -0,0 +1,58 @@
+		<meta charset="utf-8">
+		<meta name="viewport" content="width=device-width, initial-scale=1">
+		<title>jsblocks • TodoMVC</title>
+		<link rel="stylesheet" href="node_modules/todomvc-app-css/index.css">
```
Also include `todomvc-common` :)

arthurvr commented on examples/blocks/package.json:
```diff
@@ -0,0 +1,10 @@
+{
+  "name": "jsblocks-todomvc",
+  "version": "0.2.0",
+  "private": true,
```
If you're package is `private` you don't need `name` and `version`.

arthurvr commented on examples/blocks/readme.md:
```diff
@@ -0,0 +1,31 @@
+# jsblocks • [TodoMVC](http://todomvc.com)
+
+From simple user interfaces to complex single-page applications using faster, server-side rendered and easy to learn framework
+
```
Can we format this part of the readme as the other examples? (the quote on top and source + website)

arthurvr commented on examples/blocks/readme.md:
```diff
@@ -0,0 +1,31 @@
+
+## Credit
+
+Created by [Antonio Stoilkov](http://your-website.com)
```
> http://your-website.com

replace that :p

arthurvr commented:
* Please add this example to `learn.json` and to the home page.
* should the folder be named `blocks` or `jsblocks`?
* Please commit all dependencies as seen in other examples.

astoilkov commented:
No problem. I will fix everything. Can you just clarify what do you mean by adding the example to the home page? Which home page and where?

arthurvr commented:
@astoilkov [To the home page of the website, `index.html`](https://github.com/tastejs/todomvc/blob/master/index.html#L278). If you have any other questions don't hesitate to ask. Thanks for working on this!

astoilkov commented:
I think I fixed all reported things. As for the name of the folder - I have removed `js` because the npm package is called blocks and it is obvious that is a JavaScript framework. However, if you suggest that it is better to add the `js` I will rename the folder to `jsblocks`.

arthurvr commented:
> I think I fixed all reported things. As for the name of the folder - I have removed js because the npm package is called blocks and it is obvious that is a JavaScript framework. However, if you suggest that it is better to add the js I will rename the folder to jsblocks.

Well, we're not talking about an extension here. What's the name of the framework? Blocks? Or jsblocks?

arthurvr commented on examples/blocks/.gitignore:
```diff
@@ -0,0 +1,11 @@
+
+node_modules/todomvc-common/*
+!node_modules/todomvc-common/base.css
+!node_modules/todomvc-common/base.js
```
This gitignore looks okay but you forgot to actually add the files too.

astoilkov commented:
It's jsblocks. However, in jsdelivr and DefinitelyTyped the folders are called blocks. So its your call. I can't decide.

arthurvr commented on examples/blocks/index.html:
```diff
@@ -0,0 +1,59 @@
+<!doctype html>
+<html lang="en" data-framework="jsblocks">
```
this should be equal to the folder name

arthurvr commented on learn.json:
```diff
@@ -2286,6 +2286,43 @@
 			}]
 		}]
 	},
+	"blocks": {
```
this should be the same as the folder name

arthurvr commented:
Looks like I'm able to add empty todos.

arthurvr commented:
@astoilkov jsblocks, then.

arthurvr commented on examples/blocks/js/app.js:
```diff
@@ -0,0 +1,130 @@
+		},
+
+		clearCompleted: function () {
+			this.removeAll(function (todo) {
```
Might be just my feeling but `removeAll` is a confusing name if you only want to remove specific items.

astoilkov commented:
Ok. Fixed noted things. Thanks for being so patient. :)

samccone commented:
@astoilkov mind squashing down into a single commit please?

samccone commented on examples/jsblocks/index.html:
```diff
@@ -0,0 +1,59 @@
+		<link rel="stylesheet" href="node_modules/todomvc-app-css/index.css">
+	</head>
+	<body>
+		<section id="todoapp" data-query="view(Todos)" class="todoapp">
```
Should we use the v2 app spec here (classes instead of IDs) ?

astoilkov replied:
I added the id because the tests were failing without it. I can remove them quickly if this is not a problem.

arthurvr replied:
You can remove the id's. We're going to update the test suite shortly.

astoilkov commented:
When everything is perfect I will make it a single commit. Just one question. I will do it by creating a new pull request with the one commit is there any other way for making it only one commit?

arthurvr commented:
> When everything is perfect I will make it a single commit. Just one question. I will do it by creating a new pull request with the one commit is there any other way for making it only one commit?

Just squash them [using a rebase](https://help.github.com/articles/about-git-rebase/) and then force push this branch. If you've got troubles with that I'm happy to squash stuff for you while merging :)

arthurvr commented on index.html:
```diff
@@ -278,6 +278,9 @@
 									<a href="examples/webrx/" data-source="http://webrxjs.org" data-content="WebRx is a Javascript MVVM-Framework that combines functional-reactive programming with declarative Data-Binding, Templating and Client-Side Routing. The framework is built on top of ReactiveX for Javascript (RxJs) which is a powerful set of libraries for processing and querying asynchronous data-streams that can or ...
 								</li>
+								<li class="routing">
+									<a href="examples/blocks/" data-source="http://jsblocks.com" data-content="From simple user interfaces to complex single-page applications using faster, server-side rendered and easy to learn framework">jsblocks</a>
```
This link is now broken ;)

arthurvr commented:
> Looks like I'm able to add empty todos.

This is only partially fixed. I'm still able to edit an existing one with only whitespace.

arthurvr commented:
Could you please add some comments to the code? Specifically to the parts which might confuse beginners. (https://github.com/tastejs/todomvc/issues/385)

arthurvr commented on examples/jsblocks/js/app.js:
```diff
@@ -0,0 +1,130 @@
+(function () {
+	/*global blocks */
+
+	'use strict';
```
I believe it would be more consistent (and IMHO more logical) if you did

```js
/*global blocks */
(function () {
    'use strict';
```

arthurvr commented on examples/jsblocks/js/app.js:
```diff
@@ -0,0 +1,130 @@
+
+			return mode === 'active' ? !completed :
+				mode === 'completed' ? completed :
+					true;
```
This 3 lines seem like a bunch of ternary magic. Code is for humans, keep it as readable as possible :)

astoilkov replied:
Saw this from the [angular example](https://github.com/tastejs/todomvc/blob/master/examples/angularjs/js/controllers/todoCtrl.js).

arthurvr replied:
I don't like it there too :p Is there a reason to not split it up into a proper if statement?

astoilkov replied:
For me it is a little bit more clear when having a ternary operator in such cases. There are a few cases like this that I support this kind of technique. However, I agree that for begginers this could be overwhelming and switched to if statements.

arthurvr replied:
> However, I agree that for begginers this could be overwhelming and switched to if statements.

:+1:

arthurvr commented on examples/jsblocks/readme.md:
```diff
@@ -0,0 +1,27 @@
+- [Website](http://jsblocks.com)
+- [Documentation](http://jsblocks.com/learn)
+- [Download](http://jsblocks.com/download)
+- [GitHub Page](https://github.com/astoilkov/jsblocks)
```
This seems like linking up an official page 4 times. Are there any interesting blog posts we can link up?

astoilkov replied:
There is one interesting [DailyJS post](http://dailyjs.com/2015/05/29/the-jsblocks-framework/). I could add it.

arthurvr replied:
Yep, why not.

arthurvr commented:
This one starts to look realllly good @astoilkov! :+1:

astoilkov commented:
@arthurvr 

> Looks like I'm able to add empty todos.
> This is only partially fixed. I'm still able to edit an existing one with only whitespace.

I was not able to replicate that. Can you provide steps to do it?

arthurvr commented:
> I was not able to replicate that. Can you provide steps to do it?

If this helps:

1. Go to the todo app in latest Chrome
2. Clear all the todo's
3. Enter two items
4. Double click the first one to edit it
5. Clear out everything in the textbox
6. Add a bunch of whitespace (spaces)
7. Click out of the textbox to make it lose focus
8. You've got a todo item that only exists out of whitespace

This does not seem to happen every time I try it. Don't have much time ATM but will dig in. Following error in the console:

![developer_tools_-_http___localhost_3000_examples_jsblocks__and_jsblocks_ _todomvc](https://cloud.githubusercontent.com/assets/6025224/8023115/0a837c74-0cfc-11e5-8d60-7f4856534e47.png)

astoilkov commented:
@arthurvr I tried that couple of times but was not able to replicate the problem.

I commented the code, removed the ids and added the article to the readme. You could take a look at the comments and say if they are good enough.

arthurvr commented on learn.json:
```diff
@@ -2286,6 +2286,43 @@
+		"homepage": "http://jsblocks.com",
+		"examples": [{
+			"name": "Example",
+			"url": "examples/blocks"
```
`jsblocks` no?

astoilkov replied:
`jsblocks` yes.

arthurvr commented:
:star: This starts to look really good to me :star: Anyone else who wants to take a look?


> I tried that couple of times but was not able to replicate the problem.

It's strange. The tests (combined with https://github.com/tastejs/todomvc/pull/1330) doesn't seem to pass all time too but it's really rare so happy to move forward for the time being. There are many other examples that have this kind of issues.

arthurvr commented on examples/jsblocks/js/app.js:
```diff
@@ -0,0 +1,152 @@
+				include = completed;
+			}
+
+			return include;
```
Why is this variable really needed? Just return inside of the if-statements. Also you don't really need the `else if` if you return early.

astoilkov replied:
I decided to do it this way because:
 1. Further code implementation often lead to problems when returning early because other stuff that you expect to be executed does not executed...in certain situations
 2. Wwo more `return` statements in this case will lead to bigger file size
 3. In order to be more readable than the previous code I decided to add a `include` variable which hints to what the actual method purpose is and what actually means the returning of `true` and `false`

arthurvr commented:
Thanks for your neat work on this. @tastejs/todomvc anyone else who wants to take a look at this?

astoilkov commented:
Thanks for your help. The final result is awesome.

samccone commented:
On initial refresh I see a flash of raw templates, can we avoid this?

![screen shot 2015-06-17 at 10 40 53 am](https://cloud.githubusercontent.com/assets/883126/8209879/667a7602-14dd-11e5-9583-5886fdc53edc.png)

----------

Looks like this implementation has a few memory leaks that are causing the retention of DOM nodes.


##### Steps  to reproduce

* start timeline profiler
  * add three todos
  * delete three todos
  * (repeat above 2 more times)
* force GC
* stop timeline profiler

![screen shot 2015-06-17 at 10 44 04 am](https://cloud.githubusercontent.com/assets/883126/8209958/d4e89a7e-14dd-11e5-971e-7d0ea48a549d.png)
![screen shot 2015-06-17 at 10 43 33 am](https://cloud.githubusercontent.com/assets/883126/8209957/d4e5c92a-14dd-11e5-9ee2-a32da812d329.png)

astoilkov commented:
Yeah. I will make the implementation so it doesn't flicker. Regarding the memory leak I have found where the problem is. How we could proceed with this? I guess I should release a new version of the project containing the fix for the memory leak and then update the dependencies here.

samccone commented:
yes please @astoilkov, would love to see what the actual bug was here as well! Thanks for investigating

astoilkov commented:
I have fixed the problem with [this commit](https://github.com/astoilkov/jsblocks/commit/86b559d59f73d4c5eddc17414bb697990a24d631). I will test this a little more and will push a release so we could include the change in the example.

samccone commented:
:+1: thanks @astoilkov that is awesome to hear, just ping us when you are ready for another review

arthurvr commented:
This also needs a rebase to fix the merge conflict.

arthurvr commented:
ping @astoilkov, was you able to publish a release?

astoilkov commented:
No. Sorry. I was on a holiday and returned yesterday. I am a little sick now but I will publish a release next week.

arthurvr commented:
Thanks.

astoilkov commented:
Sorry for the delay. Here are the updates. :)

samccone commented:
@astoilkov looks like we need a rebase :wink:

astoilkov commented:
I am not expert on git. Is there any possibility someone else can do it? If no...can you explain me the required steps you want me to take.

sindresorhus commented:
https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

astoilkov commented:
Thanks @sindresorhus . It's time I do this.

arthurvr commented:
ping :)

astoilkov commented:
Yep. I am in process of completing it. Hope will do this in the next days.

samccone commented:
awesome thanks @astoilkov

samccone commented:
Hey @astoilkov looks like you are still failing a few tests, have you had a moment to take a look?

astoilkov commented:
Yes. I found the problem. I will try fixing it without pushing to the todomvc repository. I will write back once done.

samccone commented:
great thanks a bunch @astoilkov

astoilkov commented:
Hi @samccone. The issue is fixed. However, I will need to bump up the jsblocks version. I will fix a few more issues, bump up the version and then update it here.

samccone commented:
ok thanks for keeping us in the loop @astoilkov !!!

samccone commented:
Hey @astoilkov any word :)

astoilkov commented:
Yes. As I said I will release this week. Today we did the last bug fix that we wanted to include in the release. I will spend some time these days to make the release.

samccone commented:
awesome :sparkles: 

Just doing my weekly review of open PRs :) :)

astoilkov commented:
Ok. I am ready with the release. Now the only thing I should do is update the pull request adding the latest jsblocks version. I have a few questions though:
 - I would probably need to rebase again right? This is extremely hard for me so I ask an external person for this to help me so it could take some time.
 - I think after the last rebase we have wiped out some of the changes in the index.html and the jsblocks framework isn't there. Can you confirm that? If this is true I should add it again right?

samccone commented:
> would probably need to rebase again right? This is extremely hard for me so I ask an external person for this to help me so it could take some time.

Yep, you will want to rebase, however there are no conflicts so it should be as simple as `git fetch <remote name>`

`git rebase <remote name> master`

then `git push -f`

> If this is true I should add it again right?

Yep add it back in :)

samccone commented:
Hey guys, I am going to close the PR, just ping us when you push and think it is good to go. Thanks

astoilkov commented:
:+1:

Kanaye commented:
Hey guys. We just updated the Repo and think we are ready to go :smile:.

samccone commented:
oh looks like it freaked out :( @Kanaye mind just opening a new one and linking it? Thanks!
